### PR TITLE
new diagnostics ipynb

### PR DESCRIPTION
**Pull request by sophieeml**
new diagnostics file updated with notes from Konstantina from the previous PR - big enough changes that I created a new PR, sorry if that is not the most efficient method!

### CODE REVIEW

**Review by konstantinapanagiotopoulou — APPROVED** [review]
Thanks for updating the notebook :) It’s very helpful. My comments are a mix of small actionable points (the first one) and broader thoughts for discussion in our team meeting tomorrow (the remaining).

- **Polishing:** there are a few typos, e.g. `lables`, and most comments could be removed since the code is already pretty self-explanatory. It would also help to refine the structure with a title, numbered sections/subsections, a short purpose comment at the top, and brief interpretation sentences. Also, I think the notebook should be renamed to just `diagnostics.ipynb`. This is not that urgent, but it would be nice to revisit before submission.
- **Runtime:** it’s nice that the notebook essentially times both fitting and prediction. I’m wondering whether our runtime checks should also include preprocessing and splitting time, which might still form part of the full pipeline we need to report. Worth discussing. We should also discuss whether we need to include specific big-O notation for the from-scratch algorithm.
- **Model diagnostics:** the negative $R^2$ scores for `MHLTH` and `PHLTH` are important to flag, since they mean the model performs worse than predicting the mean. Some MSE values also seem elevated. We should probably add some interpretation of what this means and definitely mention it in the app caveats if it remains unfixed.
- **`sklearn` comparison:** `sklearn` performs extremely poorly, which is unexpected and potentially makes the comparison uninformative. Could it be due to how hyperparameters are defined between our own implementation and `sklearn`’s RBF kernel? For example, this line in Section 4 may be worth checking: `sklearn_gamma = 1 / (2 * rff_model.sigma ** 2)`.

For now, I will approve the PR so that others can also see it tomorrow! We can merge it after the meeting tomorrow.

**Comment by sophieeml** [comment]
ok everything should be updated, deleted the old diagnostics file and uploaded the new one, just called "diagnostics.ipynb"

**Review by konstantinapanagiotopoulou — COMMENTED** [review]
Thanks for the updates! A few comments:
- I noticed the notebook uses scikit-learn's KRR with an exact RBF kernel, but no RFF approximation (which is what our custom model does). We can keep this comparison to see if the approximation works, but I think we should also include a comparison to scikit-learn's KRR with an RBF kernel approximation. I looked it up and this can be done with [this tool](https://scikit-learn.org/stable/modules/generated/sklearn.kernel_approximation.RBFSampler.html#sklearn.kernel_approximation.RBFSampler) called `RBFSampler`: you can use it to approximate an RBF kernel for KRR. 
- Back to what's already done: the scikit-learn implementation does tune `gamma`, but it reuses the custom model's regularization hyperparameter for `alpha` in these lines: `skl_model = KernelRidge(alpha=rff_model.lamb, kernel="rbf", gamma=gamma)` / `sklearn_model = KernelRidge(alpha=rff_model.lamb, kernel="rbf", gamma=best_gamma)
`. That might not be the best regularization for scikit-learn's KRR, which might make the comparison unfair. 
- About runtimes: for our custom model, only prediction is timed. However, for scikit-learn, both fitting and prediction are timed. This might also make the comparison unfair and explain why our model looks much faster. Moreover, as per our meeting discussion, I think processes happening _before_ training should also be timed (e.g., `prepare_data`, `Splitter`) and compared to the whole scikit-learn pipeline. 
- There are still some "lables"/"lable" typos. 😭 Also, I think some comments can be deleted where the code is already self-explanatory, e.g., when you do the imports in the beginning or when you say "`#initialize results dict`".
- It would be nice if the `.ipynb` had a title, a short description stating its purpose, clearer markdown sectioning (`# H1: Title, ## H2: Major Section, ### H3: Subsection, #### H4: Sub-subsection`), and perhaps an interpretative sentence here and there, for example stating the takeaways from the residuals plots. 

> **The first 3 comments are more important and should be prioritized** if there's no time for all of them. The latter 2 are simpler to fix, so they can be done later.

Let me know if I can help!

**Comment by sophieeml** [comment]
thanks konstantina! the label/lable thing is tough lol can't believe I did that. Anyway i think I addressed most of these comments, reuploaded a new version! let me know what you think.

**Comment by konstantinapanagiotopoulou** [comment]
I filled in some placeholders (e.g., commentary below graphs) and made the graphs look a bit spacier to increase readability.
